### PR TITLE
Prevent users from attempting to pin PyTorch non-contiguous tensors or views only encompassing part of tensor.

### PR DESCRIPTION
**Pull request by nv-dlasalle**
## Description
This forces `pin_memory_inplace()` to go through `zerocopy_to_dgl_ndarray_for_write()` and adds checks to make sure it can be done in place, and won't pin only a subset of a tensor storage.

This does not completely address #3976 and #3800 remains an issue.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/backend/pytorch/tensor.py [hunk 1]
```diff
@@ -349,6 +349,10 @@ def zerocopy_to_dgl_ndarray(data):
         return nd.from_dlpack(dlpack.to_dlpack(data.contiguous()))
 
 def zerocopy_to_dgl_ndarray_for_write(input):
+    assert input.is_contiguous(), "Cannot convert non-contiguous tensors " \
+        "to dgl ndarray for write."
+    assert input.numel() == input.storage().size(), "Cannot convert view " \
```
Comment (yaox12) [inline python/dgl/backend/pytorch/tensor.py:354]: I doubt the error message here would confuse users. For example in this issue https://github.com/dmlc/dgl/issues/3939, users would question: "I just want to use view tensors as input, not for write. Why will it throw a 'not for write' error?"

Comment (nv-dlasalle) [reply python/dgl/backend/pytorch/tensor.py:354]: I wrapped the calls from dgl.utils with `try`/`except`, such that the user will be told pinning the memory failed in e279be8.

File: tests/pytorch/test_pin_memory.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+@unittest.skipIf(F._default_context_str == 'cpu', reason="Need gpu for this test")
+def test_pin_noncontiguous():
+    t = torch.empty([10, 100]).transpose(0, 1)
+
+    assert not t.is_contiguous()
+    assert not F.is_pinned(t)
+
+    try:
+        dgl.utils.pin_memory_inplace(t) 
+    except dgl.DGLError as e:
```
Comment (yaox12) [inline tests/pytorch/test_pin_memory.py:15]: We can use `with pytest.raises(DGLError):` context here.

**Review by jermainewang — APPROVED** [review]

File: tests/pytorch/test_pin_memory.py (hunk 2 above)
Comment (nv-dlasalle) [reply tests/pytorch/test_pin_memory.py]: Switched in 6961d66

**Pull request merged**